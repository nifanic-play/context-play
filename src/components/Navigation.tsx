import React from "react";
import { NavLink } from "react-router-dom";

export enum LABEL {
  ABOUT = "About",
  HOME = "Home",
  SERVICES = "Services",
  TEAM = "Team",
}

export interface ViewShape {
  to: string;
  label: LABEL | string;
  elements?: ViewShape[];
}

export const Navigation: React.FC = () => hierarchy(views);

const { ABOUT, HOME, SERVICES, TEAM } = LABEL;

export const views: ViewShape[] = [
  { to: "/", label: HOME },
  { to: "/services", label: SERVICES },
  {
    to: "/about",
    label: ABOUT,
    elements: [
      {
        to: "/team",
        label: TEAM,
        elements: [
          { to: "/amir", label: "Amir" },
          { to: "/nick", label: "Nick" },
        ],
      },
    ],
  },
];

export const getElementsByLabel = (label: LABEL, elements: ViewShape[] = views): ViewShape[] => {
  if (!label) throw new Error("`label` param is required");

  return elements.reduce<ViewShape[]>((acc, curr) => {
    const { elements } = curr;
    const isLabel = curr.label === label;

    if (elements && !isLabel) getElementsByLabel(label, elements);

    const { elements: extractedElements } = elements ? elements[0] : { elements: [] };

    return extractedElements || [];
  }, []);
};

/**
 * Array of parent `to` params.
 */
const historyTo: string[] = [];

const hierarchy = (items: ViewShape[], parentTo: string | undefined = "") => {
  if (!items || !items.length) return <div>Items are required to build hierarchy</div>;

  return (
    <ol>
      {items.reduce<JSX.Element[]>((acc, { to, label, elements }, i) => {
        if (elements?.length) {
          historyTo.push(to);
        }

        const item = (
          <li key={`${label}-${i}`}>
            <NavLink to={`${parentTo}${to}`}>{label}</NavLink>
            {elements?.length && hierarchy(elements, historyTo.join(""))}
          </li>
        );

        acc.push(item);

        return acc;
      }, [])}
    </ol>
  );
};
