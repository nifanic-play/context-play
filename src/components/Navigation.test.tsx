import React from "react";
import renderer from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";
import { getElementsByLabel, LABEL, Navigation, ViewShape } from "./Navigation";

describe("getElementsByLabel()", () => {
  test("returns elements of TEAM label.", () => {
    const { TEAM } = LABEL;
    const teamElements = getElementsByLabel(TEAM);
    const output: ViewShape[] = [
      { to: "/amir", label: "Amir" },
      { to: "/nick", label: "Nick" },
    ];

    expect(teamElements).toStrictEqual(output);
  });

  test("throws an error if no params provided.", () => {
    expect(getElementsByLabel).toThrow();
  });
});

describe("hierarchy()", () => {
  test("renders correctly nested `ol` navigation elements.", () => {
    const input = renderer
      .create(
        <MemoryRouter>
          <Navigation />
        </MemoryRouter>
      )
      .toJSON();

    expect(input).toMatchSnapshot();
  });
});
