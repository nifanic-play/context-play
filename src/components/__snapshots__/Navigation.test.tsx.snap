// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hierarchy() renders correctly nested \`ol\` navigation elements. 1`] = `
<ol>
  <li>
    <a
      aria-current="page"
      className="active"
      href="/"
      onClick={[Function]}
      style={Object {}}
    >
      Home
    </a>
  </li>
  <li>
    <a
      aria-current={null}
      href="/services"
      onClick={[Function]}
    >
      Services
    </a>
  </li>
  <li>
    <a
      aria-current={null}
      href="/about"
      onClick={[Function]}
    >
      About
    </a>
    <ol>
      <li>
        <a
          aria-current={null}
          href="/about/team"
          onClick={[Function]}
        >
          Team
        </a>
        <ol>
          <li>
            <a
              aria-current={null}
              href="/about/team/amir"
              onClick={[Function]}
            >
              Amir
            </a>
          </li>
          <li>
            <a
              aria-current={null}
              href="/about/team/nick"
              onClick={[Function]}
            >
              Nick
            </a>
          </li>
        </ol>
      </li>
    </ol>
  </li>
</ol>
`;
